### PR TITLE
Issue/compiler cycle breaking on functions

### LINKED ISSUE #2787: List-modified-after-freeze on specific usage of plugins
The following code causes a List-modified-after-freeze

```
import tests

   entity Rule:
     bool purged
   end

   entity Service:
     string name
   end

   Rule.services [0:] -- Service

   Service.effective_services [0:] -- Service
   Service.subservice [0:] -- Service
   implementation effective_members for Service:
        for service in subservice:
            self.effective_services += service.effective_services
        end
   end

   a = Rule(
        services = Service( name="top",
            subservice=Service(name="inner")
        ),
        purged = tests::resolve_rule_purged_status(a.services)
  )

  implementation rq for Rule:
    if self.purged:
       self.requires = services
    else:
       self.provides = services
    end
  end

  implement Rule using rq
  implement Service using effective_members
  
  ```
  
  ```
  @plugin
def resolve_rule_purged_status(
    sources: "list",
) -> "bool":
    for source in sources:
        if len(source.effective_services) == 0:
            return True

    return False

```

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Fixes a very particular bug on the compiler, where cycle breaking fails, because plugins calls (FunctionUnit) are inadvertently dropped from the all_waiters set when being re-scheduled. 

closes #2787 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [X] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] ~~End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

### CODE REVIEW

File: src/inmanta/execute/runtime.py [hunk 1]
```diff
@@ -382,7 +382,8 @@ def receive_result(self, value: ListValue, location: Location) -> None:
     def listener(self, resultcollector: ResultCollector, location: Location) -> None:
         for value in self.value:
             resultcollector.receive_result(value, location)
-        self.listeners.append(resultcollector)
```
Comment (wouterdb) [inline src/inmanta/execute/runtime.py:385]: this is a second bug triggered by the first one:

When the value is set, the listeners are dropped, causing this line to fail.

File: src/inmanta/ast/statements/call.py [hunk 2]
```diff
@@ -261,13 +262,14 @@ def __init__(self, queue_scheduler, resolver, result: ResultVariable, requires,
...
     def execute(self):
-        requires = {k: v.get_value() for (k, v) in self.requires.items()}
+        requires = {k: v.get_value() for (k, v) in self.base_requires.items()}
         try:
             self.function.resume(requires, self.resolver, self.queue_scheduler, self.result)
             self.done = True
         except UnsetException as e:
             LOGGER.debug("Unset value in python code in plugin %s %s.%s.", self.function.function, e.instance, e.attribute)
-            self.waitfor(e.get_result_variable())
+            # Don't handle it here!
```
Comment (arnaudsjs) [inline src/inmanta/ast/statements/call.py:271]: Maybe add an explanation why this shouldn't be handled here.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]
I have too little hand-on experience with the compiler code to do a quality review of this PR.

File: src/inmanta/execute/runtime.py [hunk 3]
```diff
@@ -649,6 +650,13 @@ def __init__(self, queue: QueueScheduler):
         self.queue.add_to_all(self)
         self.done = False
 
+    def _requeue_with_additional_requires(self, key: object, waitable: ResultVariable) -> None:
```
Comment (sanderr) [inline src/inmanta/execute/runtime.py:653]: Use [`typing.Hashable`](https://docs.python.org/3/library/typing.html#typing.Hashable) for `key`? It doesn't seem to reject values as strongly as you'd expect but at least it implies the intention to a user of the method.

File: tests/compiler/test_scheduling.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+from inmanta import compiler
+
+
+def test_function_rescheduling(snippetcompiler):
```
Comment (sanderr) [inline tests/compiler/test_scheduling.py:22]:
```suggestion
def test_2787_function_rescheduling(snippetcompiler):
```

File: src/inmanta/execute/runtime.py (hunk 1 above)
Comment (sanderr) [reply src/inmanta/execute/runtime.py:385]: The typing of `listeners` is very hard to follow but it doesn't look like it would be easy to fix. How comes listeners are added when the variable has already been frozen?

**Review by sanderr — CHANGES_REQUESTED** [review]
In general, could you give a bit more context on why these changes are required? I've looked at the model but I don't really understand how it fails (and it would probably take me a long time to dig it up myself). I also don't understand why the waiters need to keep track of the additional "requirements". What are those used for?

File: src/inmanta/execute/runtime.py (hunk 1 above)
Comment (wouterdb) [reply src/inmanta/execute/runtime.py:385]: I didn't dig into it very far, but: new statements being emitted (i.e. there is no reason why it should not happen, a statement can listen to something with a definite value, but then listeners must not be kept, the first half of the method feeds all values already.)

File: src/inmanta/execute/runtime.py (hunk 3 above)
Comment (wouterdb) [reply src/inmanta/execute/runtime.py:653]: ah yes, but we would have to do that all over the place (this is a good place to start), thx

File: src/inmanta/ast/statements/call.py (hunk 2 above)
Comment (wouterdb) [reply src/inmanta/ast/statements/call.py:271]: seem to have forgotten to push my changes before

File: tests/compiler/test_scheduling.py (hunk 4 above)
Comment (wouterdb) [reply tests/compiler/test_scheduling.py]: great minds, ... (I had forgotten to push this change)

**Comment by wouterdb** [comment]
> In general, could you give a bit more context on why these changes are required? I've looked at the model but I don't really understand how it fails (and it would probably take me a long time to dig it up myself). I also don't understand why the waiters need to keep track of the additional "requirements". What are those used for?

To answer these questions:

1. this model requires the `find_wait_cycle` codepath in the scheduler
2. `find_wait_cycle` requires accurate tracking of all unresolved statements (`all_waiters`)
3. `find_wait_cycle` requires accurate tracking of all ResultVariables required by these statements


- Point 2 was broken due to the premature exception catching, allowing the failed statement to be removed from `all_waiters`
- Point 3 was broken because FunctionUnits can get additional, new, requires that are discovered on the go. 
- (and the secondary bug on the listeners was also triggered by this example. 

it is basically 3 one-line fixes at once

**Review by arnaudsjs — APPROVED** [review]

File: src/inmanta/ast/statements/call.py [hunk 5]
```diff
@@ -261,13 +262,17 @@ def __init__(self, queue_scheduler, resolver, result: ResultVariable, requires,
...
+            # This exception is used by the scheduler to re-queue the unit
+            # If it is handled here, the re-queueing can not be done,
+            # leading to very subtle errors such as #2787
+            raise
```
Comment (sanderr) [inline src/inmanta/ast/statements/call.py:275]: It's still not entirely clear to me why it can't be handled here. This could just call `requeue_with_additional_requires` with the same effect, couldn't it?
(Not saying that it should, just trying to understand the subtleties of the issue.)

File: src/inmanta/execute/runtime.py [hunk 6]
```diff
@@ -649,6 +650,13 @@ def __init__(self, queue: QueueScheduler):
         self.queue.add_to_all(self)
         self.done = False
 
+    def _requeue_with_additional_requires(self, key: Hashable, waitable: ResultVariable) -> None:
```
Comment (sanderr) [inline src/inmanta/execute/runtime.py:653]: This is named as if it's meant to be a private method but it isn't. Remove the leading underscore?

File: src/inmanta/ast/statements/call.py [hunk 7]
```diff
@@ -253,6 +253,7 @@ def __init__(self, queue_scheduler, resolver, result: ResultVariable, requires,
         self.result = result
         result.set_provider(self)
         self.requires = requires
+        self.base_requires = dict(requires)
```
Comment (sanderr) [inline src/inmanta/ast/statements/call.py:256]: Perhaps add a comment here stating that `requires` is a dynamic property that is meant to keep track of all execute-time requirements while `base_requires` only contains the direct emit-time requirements.

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/ast/statements/call.py (hunk 5 above)
Comment (wouterdb) [reply src/inmanta/ast/statements/call.py:275]: No, in the `try: except` block in scheduler is another remove statement after this one. 

So catching it here allows that discard to be execute, which should not happen. 

```python
              try:
                    next.execute()
                    all_statements.discard(next)
                    count = count + 1
               except UnsetException as e:
                    # some statements don't know all their dependencies up front,...
                    next._requeue_with_additional_requires(e, e.get_result_variable())
```

Comment (sanderr) [reply src/inmanta/ast/statements/call.py:275]: I see, thanks.

File: src/inmanta/ast/statements/call.py (hunk 7 above)
Comment (wouterdb) [reply src/inmanta/ast/statements/call.py:256]: added

File: src/inmanta/execute/runtime.py (hunk 6 above)
Comment (wouterdb) [reply src/inmanta/execute/runtime.py:653]: changed

**Review by sanderr — APPROVED** [review]
